### ISSUE TITLE
filetable: fix corruption caused by internal executor behavior leading to missing constraints 

### ISSUE THREAD

**Issue by ajwerner** [cockroachdb/cockroach#76764]
**Describe the problem**

Part of the schema change protocol requires a pre-commit hook where a bunch of stuff happens related to descriptor management and safety. When you use the internal executor API with a non-nil transaction, it bypasses the lifecycle
and state management which normally happens for SQL transactions. In particular, the deferred creation of jobs and
the checking of the two version invariant.

In the case of filetable, we see just that:

https://github.com/cockroachdb/cockroach/blob/7f3bbf60c4d3f81098ab04143e3f22f571ca9199/pkg/cloud/userfile/filetable/file_table_read_writer.go#L847-L855

The intention here is good, but the outcome is dramatically bad. The end result is that the table does not have the foreign key that was intended and the descriptor is corrupt in that it references a job which was never created. 

**To Reproduce**

---

Update as of 2022-09-20: https://github.com/cockroachdb/cockroach/commit/3c44cadf4b72802ea61df2b128ab01f1baee453c fixes the bug that allowed the corruption to occur, but this issue should remain open to track work to fix any corruption that exists out in the wild.

---

```sql
CREATE DATABASE to_backup;
CREATE DATABASE backups;
BACKUP DATABASE to_backup INTO 'userfile://backups.public.userfiles_$user/data';
SELECT * FROM backups.crdb_internal.invalid_objects;
```

```
  id  | database_name | schema_name |            obj_name            |                     error
------+---------------+-------------+--------------------------------+-------------------------------------------------
  118 | backups       | public      | userfiles_$user_upload_payload | mutation job 737535981478510593: job not found
```

```sql
SELECT t->'mutationJobs', t->'mutations'
  FROM (
        SELECT crdb_internal.pb_to_json('desc', descriptor)->'table' AS t
          FROM system.descriptor
         WHERE id = '"userfiles_$user_upload_payload"'::REGCLASS
       );
```
```json
[{"jobId": "737535981478510593", "mutationId": 1}]
[
  {
    "constraint": {
      "check": {
        "constraintId": 3
      },
      "constraintType": "FOREIGN_KEY",
      "foreignKey": {
        "constraintId": 3,
        "name": "file_id_fk",
        "originColumnIds": [
          1
        ],
        "originTableId": 118,
        "referencedColumnIds": [
          2
        ],
        "referencedTableId": 117,
        "validity": "Validating"
      },
      "name": "file_id_fk",
      "uniqueWithoutIndexConstraint": {
        "constraintId": 3
      }
    },
    "direction": "ADD",
    "mutationId": 1,
    "state": "DELETE_ONLY"
  }
]
```

Jira issue: CRDB-13265

**Comment by ajwerner** [comment]
Nominally I'm assigning this bug to @cockroachdb/bulk-io but there's badness about the internal executor. I really did think we were going to do things about it this release, but guess not.

**Comment by ajwerner** [comment]
my 2c is we should consider putting out a migration while we still can for 22.1.

**Comment by dt** [comment]
This _looks_ like an internal executor bug... right?

**Comment by ajwerner** [comment]
Sure, yes, I mean, no, in that it's been this way for a long time and nobody said you could use it for schema changes and it has no contract because it's a hot pile, but also, yes, absolutely.

**Comment by ajwerner** [comment]
The way the internal executor works is that when it manages the txn lifecycle (i.e. you pass it nil), then it runs the whole state machine. When you don't you just get this weirdness where you it moves itself to the open state with totally bogus `extraTxnState` and then when the statement finishes, throws that bogus `extraTxnState` away.

**Comment by ajwerner** [comment]
https://github.com/cockroachdb/cockroach/issues/71467 and https://github.com/cockroachdb/cockroach/issues/69495 and https://cockroachlabs.slack.com/archives/C0168LW5THS/p1611947857055600 and really most closely https://cockroachlabs.slack.com/archives/CARM0PTK4/p1599052027188000 (from 2020) come really close to calling out this problem directly, but the lack of investment in doing something about those issues and the myriad other issues which come from the untamed abomination that is the internal executor is what has allowed this wound to fester.

**Comment by dt** [comment]
Ooof, so we almost certainly have some existing broken tables out there too now don't we?

Adding sql-exp as IE's owners (right?) to see how they think we should proceed here.

**Comment by mwang1026** [comment]
ping?

**Comment by rafiss** [comment]
My read above is that this issue is meant for tracking the task of fixing existing broken tables out there in the wild. I don't think SQL Experience will be picking that up.

There are already other issues for tracking how the InternalExecutor should be improved, and I agree that SQL Experience can work on those, but unfortunately I don't have a definite timeline yet for that.

**Comment by mwang1026** [comment]
ack. thanks Rafi

**Comment by rafiss** [comment]
I'm removing this from the SQL Experience board, as per my previous comment.

**Comment by rafiss** [comment]
Filed https://github.com/cockroachdb/cockroach/issues/78998 to track improvements to the transaction lifecycle management.

Though to reiterate, I'm still leaving the fix for the corruption here to the Bulk-IO team.

**Comment by ZhouXing19** [comment]
Hello, I'm trying to test the problem here with changes in #86334. Does the following means that my changes have solved this issue?

```
root@localhost:26257/defaultdb> CREATE DATABASE to_backup;
CREATE DATABASE backups;

CREATE DATABASE

CREATE DATABASE

root@localhost:26257/defaultdb> \c to_backup;
using new connection URL: postgresql://root@localhost:26257/to_backup?application_name=%24+cockroach+sql&connect_timeout=15&sslmode=disable

root@localhost:26257/to_backup> CREATE TABLE t (x int);
CREATE TABLE

root@localhost:26257/to_backup> INSERT INTO t VALUES (1), (2), (3);
INSERT 0 3

root@localhost:26257/to_backup> BACKUP DATABASE to_backup INTO 'userfile://backups.public.userfiles_$user/data';;
        job_id       |  status   | fraction_completed | rows | index_entries | bytes
---------------------+-----------+--------------------+------+---------------+--------
  792938660781752321 | succeeded |                  1 |    3 |             0 |    57
(1 row)

root@localhost:26257/to_backup> SELECT * FROM backups.crdb_internal.invalid_objects;
  id | database_name | schema_name | obj_name | error
-----+---------------+-------------+----------+--------
(0 rows)

root@localhost:26257/to_backup> \c backups;
using new connection URL: postgresql://root@localhost:26257/backups?application_name=%24+cockroach+sql&connect_timeout=15&sslmode=disable

root@localhost:26257/backups> \d
  schema_name |           table_name           | type  | owner | estimated_row_count | locality
--------------+--------------------------------+-------+-------+---------------------+-----------
  public      | userfiles_$user_upload_files   | table | root  |                   0 | NULL
  public      | userfiles_$user_upload_payload | table | root  |                   0 | NULL
(2 rows)
```

**Comment by stevendanna** [comment]
My worry about any migration we add is that it is unclear how we would know _for sure_ that the table in question was created by usertable.  The best we could do is assume based on the naming scheme.

I developed the query below to simply remove the broken constraint, which I think may be sufficient. We could then either leave it as-is or start a new schema change.  We are "safe" to drop the fk at the moment because we don't actually depend on  it for anything.

We could add this (or the go-code equivalent) as a migration; or just keep this in our back pockets for a customer who actually hits this issue.

Would appreciate some review on this query by someone from @cockroachdb/sql-schema 

```
;; The query uses unsafe_upsert_descriptor to remove a broken foreign key constraint on tables that look like
;; they are userfile related tables.
;;
;; A table is fixed iff:
;;
;; - The table name ends in _upload_payload
;; - It has exactly 1 mutationJob and 1 mutation
;; - The mutation is for the addition of a foreign key constraint in the DELETE_ONLY state.
;; - The referenced table in the foreign key has a name that ends in _upload_files
;; - The related mutationJob does not exist in the system.jobs table
;;
;; To "fix" the table, we remove the mutation and mutation job since the foreign key constrain is not strictly needed.
WITH
	possible_broken_usertables
		AS (
			SELECT
				id, (json_array_elements(d->'table'->'mutationJobs')->>'jobId')::INT8 AS job_id, d AS desc_as_json
			FROM
				(SELECT crdb_internal.pb_to_json('desc', descriptor) AS d, id FROM system.descriptor)
			WHERE
				(d->'table') IS NOT NULL
				AND d->'table'->>'name' LIKE '%_upload_payload'
				AND json_array_length(d->'table'->'mutationJobs') = 1
				AND json_array_length(d->'table'->'mutations') = 1
				AND (d->'table'->'mutations'->0->'constraint') IS NOT NULL
				AND (d->'table'->'mutations'->0->'constraint'->>'constraintType') = 'FOREIGN_KEY'
				AND (d->'table'->'mutations'->0->>'direction') = 'ADD'
				AND (d->'table'->'mutations'->0->>'state') = 'DELETE_ONLY'
				AND EXISTS(
						SELECT
							1
						FROM
							system.descriptor
						WHERE
							id = (d->'table'->'mutations'->0->'constraint'->'foreignKey'->'referencedTableId')::INT8
							AND crdb_internal.pb_to_json('desc', descriptor)->'table'->>'name' LIKE '%_upload_files'
					)
		)
SELECT
	crdb_internal.unsafe_upsert_descriptor(
		possible_broken_usertables.id,
		crdb_internal.json_to_pb(
			'cockroach.sql.sqlbase.Descriptor',
			jsonb_set(
				jsonb_set(
					jsonb_set(desc_as_json, ARRAY['table', 'mutationJobs'], '[]'),
					ARRAY['table', 'mutations'],
					'[]'
				),
				ARRAY['table', 'version'],
				((desc_as_json->'table'->>'version')::INT8 + 1)::STRING::JSONB
			)
		)
	)
FROM
	possible_broken_usertables LEFT JOIN system.jobs ON possible_broken_usertables.job_id = system.jobs.id
WHERE
	system.jobs.id IS NULL;
```

**Comment by stevendanna** [comment]
Fixed in aaaea8b991391d355bd52d6cba6cf7149ab9fc50
